File: kdcreader_android/android/src/main/kotlin/koamtac/kdc/sdk/plugins/view/KComponentView.kt
package koamtac.kdc.sdk.plugins.view

import android.content.Context
import android.view.View
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.platform.PlatformView
import koamtac.kdc.sdk.KDCSWDecoderComponent

class KComponentView(context: Context?, id: Int, messenger: BinaryMessenger, params: Map<*, *>?) :
    PlatformView, MethodCallHandler {

    companion object {
        private const val SW_DECODER_COMPONENT_VIEW_METHOD_CHANNEL_NAME = "plugins.flutter.koamtac/swcomponentview"

        private const val ENABLE_SCANNING = "enableScanning"
        private const val SET_FREEZE_MODE = "setFreezeMode"

        private const val PARAM_NAME_SCANNING = "scanning"
        private const val PARAM_NAME_FREEZE_MODE = "freeze"
    }

    private val componentView: KDCSWDecoderComponent = KDCSWDecoderComponent(context)
    private val methodChannel : MethodChannel

    override fun getView(): View {
        return componentView
    }

    override fun dispose() {
        methodChannel.setMethodCallHandler(null)
    }

    init {
//        val scanning: Boolean? = params?.get(PARAM_NAME_SCANNING) as Boolean?
//        scanning?.let {
//            componentView.enableScanning(it)
//        }
//
//        val freeze: Boolean? = params?.get(PARAM_NAME_FREEZE_MODE) as Boolean?
//        freeze?.let {
//            componentView.setFreezeMode(it)
//        }

        methodChannel = MethodChannel(messenger, SW_DECODER_COMPONENT_VIEW_METHOD_CHANNEL_NAME + "_$id")
        methodChannel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            ENABLE_SCANNING -> enableScanning(call, result)
            SET_FREEZE_MODE -> setFreezeMode(call, result)
            else -> result.notImplemented()
        }
    }

    private fun enableScanning(call: MethodCall, result: MethodChannel.Result) {
        val isEnable = call.arguments<Boolean>()
        isEnable?.let {
            componentView.enableScanning(it)
        }
    }

    private fun setFreezeMode(call: MethodCall, result: MethodChannel.Result) {
        val isEnable = call.arguments<Boolean>()
        isEnable?.let {
            componentView.setFreezeMode(it)
        }
    }
}

File: kdcreader_ios/ios/Classes/view/KComponentView.swift
//
//  KComponentView.swift
//  device_info_plus
//
//  Created by koamtac on 4/3/24.
//

import Flutter

class KComponentView: NSObject, FlutterPlatformView {
    private let SW_DECODER_COMPONENT_VIEW_METHOD_CHANNEL_NAME = "plugins.flutter.koamtac/swcomponentview"
    
    private let ENABLE_SCANNING = "enableScanning"
    private let SET_FREEZE_MDOE = "setFreezeMode"
    
    /* PARAM */
    private let PARAM_NAME_SCANNING = "scanning"
    private let PARAM_NAME_FREEZE_MODE = "freeze"
    
    private var _view: UIView
    
    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?,
        binaryMessenger messenger: FlutterBinaryMessenger?
    ) {
        _view = KNativeView(frame: frame)

        super.init()

// TBD
//        if let nativeView = _view as? KNativeView,
//           let params = args as? Dictionary<String, Any> {
//            if let enable = params[PARAM_NAME_SCANNING] as? Bool {
//                nativeView.componentView?.enable(enable)
//            }
//         
//            if let enable = params[PARAM_NAME_FREEZE_MODE] as? Bool {
//                nativeView.componentView?.setFreezeMode(enable)
//            }
//        }
        
        if let m = messenger {
            initMethodChannel(viewId, messenger: m)
        }
    }

    func view() -> UIView {
        return _view
    }

    private func initMethodChannel(_ viewId: Int64, messenger: FlutterBinaryMessenger) {
        let methodChannel = FlutterMethodChannel(name: SW_DECODER_COMPONENT_VIEW_METHOD_CHANNEL_NAME + "_\(viewId)", binaryMessenger: messenger)
        methodChannel.setMethodCallHandler { call, result in
            switch call.method {
            case self.ENABLE_SCANNING:
                self.enableScanning(call, result: result)
                
            case self.SET_FREEZE_MDOE:
                self.setFreezeMode(call, result: result)
                
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
    
    private func enableScanning(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let enable = call.arguments as? Bool else {
            return
        }
        
        if let nativeView = _view as? KNativeView {
            nativeView.componentView?.enable(enable)
        }
    }
    
    private func setFreezeMode(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let enable = call.arguments as? Bool else {
            return
        }

        if let nativeView = _view as? KNativeView {
            nativeView.componentView?.setFreezeMode(enable)
        }
    }
}

class KNativeView: UIView {
    var componentView: HSMDecodeComponent?
    
    override init(frame: CGRect) {
        componentView = HSMDecodeComponent(frame: frame)
        super.init(frame: frame)

        initComponentView()
    }
    
    required init?(coder: NSCoder) {
        componentView = HSMDecodeComponent(coder: coder) ?? nil
        super.init(coder: coder)
        
        initComponentView()
    }
    
    override func layoutSubviews() {
        if let child = componentView {
            // Re-enable scan to fit the camera screen to the view size
            if child.isEnabled() {
                child.enable(false)
                child.enable(true)
            }
        }
        super.layoutSubviews()
    }
    
    private func initComponentView() {
        if let child = componentView {
            self.addSubview(child)
            
            child.translatesAutoresizingMaskIntoConstraints = false
            
            let con = [
                child.centerXAnchor.constraint(equalTo: centerXAnchor),
                child.centerYAnchor.constraint(equalTo: centerYAnchor),
                child.widthAnchor.constraint(equalTo: widthAnchor),
                child.heightAnchor.constraint(equalTo: heightAnchor)
            ]
            NSLayoutConstraint.activate(con)
            
            // To show the camera screen when it starts.
            child.enable(true)
        }
    }
}
